qualtran/bloqs/for_testing/random_select_and_prepare.py

```diff
@@ -0,0 +1,157 @@
+#  Copyright 2024 Google LLC
+#
+#  Licensed under the Apache License, Version 2.0 (the "License");
+#  you may not use this file except in compliance with the License.
+#  You may obtain a copy of the License at
+#
+#      https://www.apache.org/licenses/LICENSE-2.0
+#
+#  Unless required by applicable law or agreed to in writing, software
+#  distributed under the License is distributed on an "AS IS" BASIS,
+#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#  See the License for the specific language governing permissions and
+#  limitations under the License.
+from functools import cached_property
+from typing import Optional, Sequence, Tuple
+
+import attrs
+import cirq
+import numpy as np
+from attrs import frozen
+from numpy.typing import NDArray
+
+from qualtran import BloqBuilder, BoundedQUInt, QBit, Register, SoquetT
+from qualtran.bloqs.for_testing.random_gate import RandomGate
+from qualtran.bloqs.qubitization_walk_operator import QubitizationWalkOperator
+from qualtran.bloqs.select_and_prepare import PrepareOracle, SelectOracle
+
+
+@frozen
+class RandomPrepareOracle(PrepareOracle):
+    U: RandomGate
+
+    @property
+    def selection_registers(self) -> tuple[Register, ...]:
+        return (Register('selection', BoundedQUInt(bitsize=self.U.bitsize)),)
+
+    @staticmethod
+    def create(bitsize: int, *, random_state: np.random.RandomState):
+        matrix = RandomGate.create(bitsize, random_state=random_state).matrix
+        matrix = np.array(matrix)
+
+        # make the first column (weights alpha_i) all reals
+        alpha = matrix[:, 0]
+        matrix = matrix * (alpha.conj() / np.abs(alpha))[:, None]
+
+        # verify that it is still unitary
+        np.testing.assert_allclose(matrix @ matrix.conj().T, np.eye(2**bitsize), atol=1e-10)
+        np.testing.assert_allclose(matrix.conj().T @ matrix, np.eye(2**bitsize), atol=1e-10)
+
+        return RandomPrepareOracle(RandomGate(bitsize, matrix))
+
+    def build_composite_bloq(self, bb: BloqBuilder, selection: SoquetT) -> dict[str, SoquetT]:
+        selection = bb.add(self.U, q=selection)
+        return {'selection': selection}
+
+    def __pow__(self, power):
+        if power == -1:
+            return self.U.adjoint()
+        return NotImplemented
+
+    @cached_property
+    def alphas(self):
+        column = np.array(self.U.matrix)[:, 0]
+        np.testing.assert_almost_equal(np.imag(column), 0)
+        return column**2
+
+
+@frozen
+class PauliSelectOracle(SelectOracle):
```

Comment:
Please add a docstring to the class. Specify things like number of qubits it needs and whether it allocates additional qubits as part of the decomposition.